### PR TITLE
fix: remove dependence @types/webpack and fix type error

### PR DESCRIPTION
**Pull request by tqma113**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
<details>
<summary>
Error log
</summary>
<pre>
<code>
> definitely-typed@0.0.3 test /Users/matianqi/Projects/JavaScript/DefinitelyTyped
> dtslint types "webpack-dev-middleware"

Error: Errors in typescript@4.2 for external dependencies:
../uglify-js/index.d.ts(9,30): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`
../webpack-sources/lib/CachedSource.d.ts(1,30): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`
../webpack-sources/lib/ConcatSource.d.ts(2,28): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`
../webpack-sources/lib/index.d.ts(1,30): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`
../webpack-sources/lib/OriginalSource.d.ts(2,28): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`
../webpack-sources/lib/Source.d.ts(2,30): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`
../webpack-sources/lib/SourceMapSource.d.ts(1,50): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`
../webpack/index.d.ts(46,30): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`

    at /Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/dtslint/bin/index.js:206:19
    at Generator.next (<anonymous>)
    at fulfilled (/Users/matianqi/Projects/JavaScript/DefinitelyTyped/node_modules/dtslint/bin/index.js:6:58)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! definitely-typed@0.0.3 test: `dtslint types "webpack-dev-middleware"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the definitely-typed@0.0.3 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/matianqi/.npm/_logs/2021-01-24T17_44_40_629Z-debug.log
</code>
</pre>
</details>

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/DefinitelyTyped/DefinitelyTyped/issues/50773>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by tqma113** [comment]
#50773

**Comment by peterblazejewicz** [comment]
So the error is due to resolving (via '../' in tsconfig) into v4 @types/webpack, where 'watching' is not a property of an interface/class.
This can be fixed by adding types resolution in paths IMO, let me ask for help with this (unless you're sure how to do this)

**Review by peterblazejewicz — APPROVED** [review]
LGMT!
ready to merge once we fix type for v5 resolution

**Comment by peterblazejewicz** [comment]
@tqma113 is this supposed to be resolved against v5 of webpack? If so, there should be `package.json` added with explicit dependency entry.

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]
Should we add `package.json`?

**Comment by tqma113** [comment]
@peterblazejewicz  If add the `"webpack":"^5.0.0"` to `dependences` of `package.json`, when I install the package there may be a duplicate `webpack` package. The better way is `peerDependences`, but this has no effect on the issue we are currently experiencing.

**Comment by peterblazejewicz** [comment]
I was thinking about `tsconfig`, but obviously it will work for DT package here only during CI tests.
For consuming types, I'd guess same entry should be added, e.g. how to resolve 'webpack' types, to avoid `@types/webpack` and native definition clashes in case like that one. This compatiblity layer in 4.* version does not cover your PR scenario, sadly. Let me try, tbu

**Comment by slavafomin** [comment]
I was trying to replace `@types/webpack` with `webpack` in `webpack-dev-server` and encountered a similar issue: `Could not find a declaration file for module 'source-map'`. Have you managed to fix it?

I also believe that webpack should got to a `peerDependencies` instead of the `dependencies`, but I'm not sure if it's possible with DefinitelyTyped infrastructure.

File: types/webpack-dev-middleware/index.d.ts [hunk 1]
```diff
@@ -66,7 +66,7 @@ declare namespace WebpackDevMiddleware {
         callbacks: Callback[];
         options: Options;
         compiler: webpack.Compiler;
-        watching: webpack.Watching | null;
+        watching: webpack.Compiler["watching"] | null;
```
Comment (unstubbable) [inline types/webpack-dev-middleware/index.d.ts:69]: With the latest version of Webpack this is not needed anymore, `webpack.Watching` is now available.

**Comment by peterblazejewicz** [comment]
@tqma113 i've meant `package.json` on DT package level, so `types/webpack-dev-middleware/package.json`.
The repo root package should not be changed at all.

**Pull request closed without merging**